src/contrib/dependencyinjection/Akka.DependencyInjection/ServiceProvider.cs

```diff
@@ -0,0 +1,145 @@
+        /// </remarks>
+        /// <typeparam name="T">The type of actor to instantiate.</typeparam>
+        /// <param name="args">Optional. Any constructor arguments that will be passed into the actor's constructor directly without being resolved by DI first.</param>
+        /// <returns>A new <see cref="Akka.Actor.Props"/> instance which uses DI internally.</returns>
+        public Props Props<T>(params object[] args) where T : ActorBase
+        {
+            return new ServiceProviderProps<T>(Provider, args);
+        }
+    }
+
+    /// <summary>
+    /// INTERNAL API
+    /// </summary>
+    public sealed class ServiceProviderExtension : ExtensionIdProvider<ServiceProvider>
+    {
+        public override ServiceProvider CreateExtension(ExtendedActorSystem system)
+        {
+            var setup = system.Settings.Setup.Get<ServiceProviderSetup>();
+            if (!setup.HasValue)
+            {
+                var exception = new ConfigurationException("Unable to find [ServiceProviderSetup] included in ActorSystem settings." +
+                                                           " Please specify one before attempting to use dependency injection inside Akka.NET.");
+                system.EventStream.Publish(new Error(exception, "Akka.DependencyInjection", typeof(ServiceProviderExtension), exception.Message));
+                throw exception;
+            }
+
+            return new ServiceProvider(setup.Value.ServiceProvider);
+        }
+    }
+
+    /// <summary>
+    /// This class represents a specialized <see cref="Akka.Actor.Props"/> that uses delegate invocation
+    /// to create new actor instances, rather than a traditional <see cref="System.Activator"/>.
+    ///
+    /// Relies on having an active <see cref="IServiceProvider"/> implementation available
+    /// </summary>
+    /// <typeparam name="TActor">The type of the actor to create.</typeparam>
+    internal class ServiceProviderProps<TActor> : Props where TActor : ActorBase
+    {
+        private readonly IServiceProvider _provider;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ServiceProviderProps{TActor}" /> class.
+        /// </summary>
+        /// <param name="provider">The <see cref="IServiceProvider"/> used to power this class</param>
+        /// <param name="args">The constructor arguments passed to the actor's constructor.</param>
+        public ServiceProviderProps(IServiceProvider provider, params object[] args)
+            : base(typeof(TActor), args)
+        {
+            _provider = provider;
+        }
+
+        /// <summary>
+        /// Creates a new actor using the configured factory method.
+        /// </summary>
+        /// <returns>The actor created using the factory method.</returns>
+        public override ActorBase NewActor()
+        {
+            return ActivatorUtilities.CreateInstance<TActor>(_provider, Arguments);
```

Comment:
Removed the `Func<IServiceProvider, TActor>` delegate here due to the memory leak point that @Zetanova raised. This is much cleaner.